rustc_target: introduce Abi, Env, Os

tamird opened the pull request:
Improve type safety by using an enum rather than strings.

I'm not really sure this is better since only a few vendors have special semantics. r? @nnethercote

RalfJung commented:
FWIW for a related change recently we had quite a bit of discussion: https://github.com/rust-lang/compiler-team/issues/926.

tamird committed:
rustc_target: introduce Abi

Improve type safety by using an enum rather than strings.

tamird committed:
rustc_target: introduce Env

Improve type safety by using an enum rather than strings.

tamird committed:
rustc_target: introduce Os

Improve type safety by using an enum rather than strings.

nnethercote commented on compiler/rustc_target/src/spec/mod.rs:
```diff
@@ -1930,6 +1930,44 @@ impl Arch {
+        Win7 = "win7",
+        Wrs = "wrs",
+    }
+    /// The vast majority of the time, the compiler deals with a fixed set of
```
This comment is repeated multiple times with only slight variations. I think it would be fine to instead say "// See `enum Target` for why this `Unknown` variant exists."  (The `OS` type has an extra sentence that can be kept.)

tamird replied:
Moved, thanks.

nnethercote commented on compiler/rustc_target/src/spec/mod.rs:
```diff
@@ -1930,6 +1930,44 @@ impl Arch {
+        Unikraft = "unikraft",
+        Uwp = "uwp",
+        Vex = "vex",
+        Win7 = "win7",
```
Any idea what "win7" is? Makes me think of Windows 7, which is a weird name for a vendor.

nnethercote replied:
Ugh, ok. Thanks for the info!

nnethercote commented on compiler/rustc_target/src/spec/mod.rs:
```diff
@@ -2060,7 +2098,7 @@ pub struct TargetOptions {
     /// However, parts of the backend do check this field for specific values to enable special behavior.
     pub abi: StaticCow<str>,
     /// Vendor name to use for conditional compilation (`target_vendor`). Defaults to "unknown".
```
The comment needs updating.

nnethercote commented on compiler/rustc_target/src/spec/mod.rs:
```diff
@@ -1930,6 +1930,77 @@ impl Arch {
+        Windows = "windows",
+        Xous = "xous",
+        Zkvm = "zkvm",
+        Ios = "ios",
```
Move this up to be in alphabetical order. Also, for consistency it should be `IOs`, as ugly as that looks.

nnethercote reviewed (changes requested):
I have a bunch of nits but this generally looks good. `Linker` is another candidate for enum-ification, if you aren't yet bored of this stuff :)

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/148531/401ae55427522984e4a89c37cff6562a4ddcf6b7..dbd3aa21fa5971b510ea394b3574f54ef25c6f9b)*

tamird committed:
rustc_target: rehome target_abi comment

tamird commented on compiler/rustc_target/src/spec/mod.rs:
```diff
@@ -1930,6 +1930,77 @@ impl Arch {
+        NetBsd = "netbsd",
+        None = "none",
+        Nto = "nto",
+        Nuttx = "nuttx",
```
If you say so! Done.

tamird commented on compiler/rustc_target/src/spec/mod.rs:
```diff
@@ -1930,6 +1930,77 @@ impl Arch {
+        SolidAsp3 = "solid_asp3",
+        TeeOs = "teeos",
+        Trusty = "trusty",
+        Tvos = "tvos",
```
Done.

tamird commented on compiler/rustc_codegen_cranelift/src/lib.rs:
```diff
@@ -47,7 +47,7 @@ use rustc_middle::dep_graph::{WorkProduct, WorkProductId};
 use rustc_session::config::OutputFilenames;
 use rustc_span::{Symbol, sym};
-use rustc_target::spec::Arch;
+use rustc_target::spec::{ABI, Arch};
```
Done.

tamird committed:
rustc_target: move comment to macro

Avoid duplicating this comment in preparation for adding more enums.

tamird reviewed (commented):
Addressed all but one comment and left things unsquashed to (somewhat) ease review. I do plan to squash once you're happy with it.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/148531/c90bcb9571b7aab0d8beaa2ce8a998ffaf079d38..367f96eb8beaf598eed0f421abcd6c5980f8c238)*

nnethercote commented on compiler/rustc_target/src/lib.rs:
```diff
@@ -135,6 +135,14 @@ macro_rules! target_spec_enum {
+            /// an enum. However, it is possible to have arbitrary values in a
+            /// target JSON file (which can be parsed when `--target` is
+            /// specified). This might occur, for example, for an out-of-tree
+            /// codegen backend that supports a value that rustc currently
```
Can you change "value" to "value (e.g. architecture or OS)". I think it's helpful to have concrete examples.

tamird replied:
Done.

nnethercote commented on compiler/rustc_target/src/spec/mod.rs:
```diff
@@ -2552,7 +2590,7 @@ impl Default for TargetOptions {
             os: "none".into(),
             env: "".into(),
             abi: "".into(),
-            vendor: "unknown".into(),
```
Probably. Though I don't want to make you change it for `Arch` when I already made you change it the other way.

So:
- having `Os::Unknown` (which is the default), `Os::None`, and `Os::Other(String)` -- yes please.
- changing `{Arch,Env,Vendor,Abi}::Unknown(String)`  to `{Arch,Env,Vendor,Abi}::Other(String)` -- if you can be bothered.

tamird committed:
rustc_target: rename Arch::{Uknown,Other}

Prepare for additional enums like Vendor and Os which have true
`Unknown` variants. We want to use the same name for the escape hatch
for all of these, thus rename this one.

tamird commented:
range-diff is ~empty, which is what I expected!

@bors r=@nnethercote

bors commented:
:pushpin: Commit cd47df7c7501b5e37c6c91deae14937ff0d6a50e has been approved by `nnethercote`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=nnethercote cd47df7c7501b5e37c6c91deae14937ff0d6a50e -->
<!-- homu: {"type":"Approved","sha":"cd47df7c7501b5e37c6c91deae14937ff0d6a50e","approver":"nnethercote","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

tamird committed:
rustc_codegen_ssa: reduce repetition

bors commented:
:umbrella: The latest upstream changes (presumably #147935) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

madsmtm commented on compiler/rustc_target/src/spec/mod.rs:
```diff
@@ -1928,11 +1921,167 @@ impl Arch {
+}
+
+crate::target_spec_enum! {
+    pub enum Vendor {
```
Hmm, I'm semi against making `target_vendor` an enum - it's not like the other ones, in that it's not supposed to carry any meaning (at least not after https://github.com/rust-lang/lang-team/issues/102).

I'd prefer to keep it as just a `String`.

tamird replied:
That sounds nice, but it is just not the case. There are many places where the vendor is consulted, especially to check if it's `Apple`.

tamird replied:
```
% git grep -E '[^:] Vendor::' -- ':!compiler/rustc_target/src/spec/targets/' ':!compiler/rustc_target/src/spec/base'
compiler/rustc_codegen_cranelift/src/abi/mod.rs:            (Arch::X86_64, _) | (Arch::AArch64, Vendor::Apple) => match (ty, is_signed) {
compiler/rustc_codegen_cranelift/src/codegen_f16_f128.rs:        if fx.tcx.sess.target.vendor == Vendor::Apple && fx.tcx.sess.target.arch == Arch::X86_64 {
compiler/rustc_codegen_cranelift/src/codegen_f16_f128.rs:        if fx.tcx.sess.target.vendor == Vendor::Apple && fx.tcx.sess.target.arch == Arch::X86_64 {
compiler/rustc_codegen_cranelift/src/codegen_f16_f128.rs:        if fx.tcx.sess.target.vendor == Vendor::Apple && fx.tcx.sess.target.arch == Arch::X86_64 {
compiler/rustc_codegen_ssa/src/back/link.rs:                        && sess.target.vendor != Vendor::Uwp
compiler/rustc_codegen_ssa/src/back/linker.rs:    if matches!(flavor, LinkerFlavor::Msvc(..)) && t.vendor == Vendor::Uwp {
compiler/rustc_codegen_ssa/src/back/linker.rs:    assert!(cmd.get_args().is_empty() || sess.target.vendor == Vendor::Uwp);
compiler/rustc_codegen_ssa/src/common.rs:    target.vendor == Vendor::Pc
compiler/rustc_metadata/src/native_libs.rs:    if sess.target.vendor == Vendor::Fortanix
compiler/rustc_target/src/spec/mod.rs:            self.vendor == Vendor::Apple,
compiler/rustc_target/src/spec/mod.rs:        if let Vendor::Other(s) = &self.vendor {
compiler/rustc_target/src/spec/mod.rs:            || (self.env == Env::Sgx && self.vendor == Vendor::Fortanix)
src/tools/miri/src/machine.rs:                    if target.options.vendor == Vendor::Apple {
```
It's not just Apple, either.

tamird commented:
@bors r=@nnethercote

bors commented:
:pushpin: Commit 6b87b8788efe9c2c446ed9562ad233e454819284 has been approved by `nnethercote`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=nnethercote 6b87b8788efe9c2c446ed9562ad233e454819284 -->
<!-- homu: {"type":"Approved","sha":"6b87b8788efe9c2c446ed9562ad233e454819284","approver":"nnethercote","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bjorn3 commented on compiler/rustc_target/src/spec/mod.rs:
```diff
@@ -1928,11 +1921,167 @@ impl Arch {
+        Sony = "sony",
+        Sun = "sun",
+        Unikraft = "unikraft",
+        Unknown = "unknown",
```
This should be placed under `Other` IMO. Matching on `Vendor::Unknown` never makes sense under any circumstance. It is a catch-all for when there is no known vendor value, which is also what the `Other` variant is for.

madsmtm commented:
I'm gonna unmark this as accepted, since I'd like to see the `Vendor` changes reverted, and that we go with the plan discussed in https://github.com/rust-lang/rust/pull/148531#discussion_r2507383167 instead (preferably making those changes in a separate PR).
@bors r-

(Do correct me if this is bad form).

We should bump the priority next time it's scheduled for merging, to avoid further merge conflicts.

nnethercote commented:
@tamird: looks like you are planning to merge #148760 before this, and then adjust this accordingly? I give you a gold star for flexibility. It's annoying to do a nice refactoring cleanup that follows the contours of the existing code and then be told that the existing code shouldn't exist. ⭐

tamird commented:
> @tamird: looks like you are planning to merge #148760 before this, and then adjust this accordingly? I give you a gold star for flexibility. It's annoying to do a nice refactoring cleanup that follows the contours of the existing code and then be told that the existing code shouldn't exist. ⭐

Annoying is a crucial part of doing OSS :)

bors commented:
:pushpin: Commit 98a534e1fa35ef74aafee60f8c5bcb40bbcc1efb has been approved by `nnethercote`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=nnethercote 98a534e1fa35ef74aafee60f8c5bcb40bbcc1efb -->
<!-- homu: {"type":"Approved","sha":"98a534e1fa35ef74aafee60f8c5bcb40bbcc1efb","approver":"nnethercote","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 98a534e1fa35ef74aafee60f8c5bcb40bbcc1efb with merge 0b329f801a09004dacb19aaf09d5cb8b4c51d3f8...
<!-- homu: {"type":"BuildStarted","head_sha":"98a534e1fa35ef74aafee60f8c5bcb40bbcc1efb","merge_sha":"0b329f801a09004dacb19aaf09d5cb8b4c51d3f8"} -->

tamird commented on compiler/rustc_target/src/spec/mod.rs:
```diff
@@ -1928,11 +1921,135 @@ impl Arch {
+}
+
+crate::target_spec_enum! {
+    pub enum Abi {
```
Certainly agree. This was a mechanical change.
